modules/test/playwright/tests/client-extension-web/pages/PageEditorPage.ts

```diff
@@ -0,0 +1,73 @@
+/**
+ * SPDX-FileCopyrightText: (c) 2000 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+import {Page, expect} from '@playwright/test';
+
+import {Layout} from '../../../helpers/json-web-services/JSONWebServicesLayoutApiHelper';
+import {liferayConfig} from '../../../liferay.config';
+
+export class PageEditorPage {
+	readonly page: Page;
+	readonly layout: Layout;
+
+	private editMode: boolean;
+
+	constructor(page: Page, layout: Layout) {
+		this.page = page;
+		this.layout = layout;
+		this.editMode = false;
+	}
+
+	async addWidget(category: string, name: string) {
+		this.assertEditMode(true);
+
+		await this.page
+			.getByRole('tab', {exact: true, name: 'Widgets'})
+			.click();
+		await this.page.getByRole('menuitem', {name: category}).click();
+		await this.page.getByText(name).click();
+		await this.page
+			.getByRole('menuitem', {
+				name: `${name} Add ${name} Mark ${name} as Favorite`,
+			})
+			.press('Tab');
+		await this.page.getByLabel(`Add ${name}`).press('Enter');
+		await this.page.getByLabel(`Add ${name}`).press('Enter');
+	}
```

Comment:
Can we add widgets through headless API, like [here](https://github.com/liferay/liferay-portal/blob/master/modules/test/playwright/tests/client-extension-web/editorConfigContributor.spec.ts#L101-L112)?

Reply:
Ah. Didn't know that was possible. :open_mouth: 

However, does that create a widget or a content page? :thinking: Looks like it works for content pages... I will give it a try.

Reply:
I've asked Echo and adding a widget to a fragment page via headless is not yet possible. The only possibility is what you are doing in the linked code: create a site with a page in one shot. However, I think that's too much in this case, because the suite would create a new site per test which is way too much overhead, IMO, for little benefit (since the UI code is so small and it's centralized).